File: 503-next-greater-element-ii/503-next-greater-element-ii.py
class Solution:
    def monotonicstack(self, arr, n):
        stack = []  # monotonic stack
        nge = [0] * n
        for i in range(2 * n - 1, -1, -1):
            while stack != [] and arr[i % n] >= arr[stack[-1]]:
                stack.pop()
            if stack == []:
                nge[i % n] = -1
            else:
                nge[i % n] = stack[-1]
            stack.append(i % n)
        return nge
    
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        greater = self.monotonicstack(nums, len(nums))
        ans = [nums[greater[i]] if greater[i] != -1 else -1 for i in range(len(nums))]
        return ans

File: 503-next-greater-element-ii/503-next-greater-element-ii.cpp
class Solution {
public:
    
    vector<int> monotonicStack(vector<int> arr, int n) {
        stack<int> stack;
        vector<int> nge(n, -1);
        for (int i = 2 * n - 1; i > -1; i--) {
            while ((stack.size() != 0) and (arr[i % n] >= stack.top())) {
                stack.pop();
            }
            if (stack.size() != 0) {
                nge[i % n] = stack.top();
            }
            stack.push(arr[i % n]);
        }
        return nge;
    }
    vector<int> nextGreaterElements(vector<int>& nums) {
        vector<int> greater = monotonicStack(nums, nums.size());
        return greater;
    }
};
